quickwit-storage/src/cache/disk/cache.rs

```diff
@@ -0,0 +1,330 @@
+        } else {
+            self.cache_counters.misses_num_items.inc();
+        }
+    }
+
+    fn calculate_file_to_evict(&self, num_bytes: usize) -> Vec<FileKey> {
+        if self.capacity.exceeds_capacity(num_bytes) {
+            // The value does not fit in the cache. We simply don't store it.
+            warn!(
+                capacity_in_bytes = ?self.capacity,
+                len = num_bytes,
+                "Downloaded a byte slice larger than the cache capacity."
+            );
+            return vec![];
+        }
+
+        let current_num_bytes = self.num_bytes.load(Ordering::Relaxed);
+
+        let mut to_clear = vec![];
+        let mut lru_lock = self.lru.lock();
+        while self
+            .capacity
+            .exceeds_capacity(current_num_bytes as usize + num_bytes)
+        {
+            let maybe_file_metadata = lru_lock
+                .peek_lru()
+                .and_then(|(k, _)| self.store.get_metadata(*k));
+
+            if let Some(metadata) = maybe_file_metadata {
+                let time_delta = super::time_now() - Duration::from_secs(metadata.last_accessed);
+                if time_delta < MIN_TIME_SINCE_LAST_ACCESS {
+                    // Reset the lru state.
+                    reset_lru_state(&mut lru_lock, to_clear);
+
+                    // It is not worth doing an eviction.
+                    return vec![];
+                }
+            }
+
+            if let Some((key, _)) = lru_lock.pop_lru() {
+                to_clear.push(key);
+            } else {
+                reset_lru_state(&mut lru_lock, to_clear);
+                error!(
+                    "Logical error. Even after removing all of the items in the cache the \
+                     capacity is insufficient. This case is guarded against and should never \
+                     happen."
+                );
+                return vec![];
+            }
+        }
+
+        to_clear
+    }
+
+    async fn check_and_evict(&self, num_bytes: usize) -> bool {
+        let to_clear = self.calculate_file_to_evict(num_bytes);
+        if to_clear.is_empty() {
+            return false;
+        }
+
+        let store = self.store.clone();
+
+        let to_remove = {
+            let mut old_files = self.retry_files.lock().clone();
+            old_files.extend(to_clear);
+            old_files
+        };
+        let to_remove_copy = to_remove.clone();
+
+        let result = tokio::task::spawn_blocking(move || {
+            let mut failed_files = vec![];
+            for key in to_remove {
+                if let Err(e) = store.remove_with_key(key) {
+                    warn!(
+                        error = ?e,
+                        "IO Error: Failed to remove file from disk cache. A retry will happen on the next write.",
+                    );
+                    failed_files.push(key);
+                };
+            }
+
+            failed_files
+        }).await;
```

Comment:
It seemed that we don't want to necessarily abort writing a file to the cache if another file can't be removed for some reason. Instead, the system just marks the file as something to attempt to remove on the next write. Extending this would probably add a backoff system or some method of tracking how many times a file has attempted to be removed.